packages/@aws-cdk/cdk/lib/physical-name-generator.ts

```diff
@@ -0,0 +1,81 @@
+import crypto = require('crypto');
+import { IResource } from "./resource";
+import { Stack } from './stack';
+
+export class PhysicalNameGenerator {
+  public static from(resource: IResource): PhysicalNameGenerator {
+    return new PhysicalNameGenerator(
+        new PrefixNamePart(resource.node.stack.name, 25),
+        new SuffixNamePart(resource.node.uniqueId, 24),
+        resource.node.stack);
+  }
+
+  private readonly stackPart: NamePart;
+  private readonly idPart: NamePart;
+  private readonly stack: Stack;
+  private readonly hashLength: number;
+  private changeToLowerCase = false;
+
+  private constructor(stackPart: NamePart, idPart: NamePart, stack: Stack) {
```

Comment:
Why not just pass the `IResource` and make all these `NamePart` things internal/private implementation details.

Reply:
That's exactly what is happening (the constructor is private, the creation of the generator happens through the static factory method `from`, which takes an `IResource` as the only argument).